StorageUtility/Linux/StorageUtility.cpp

```diff
@@ -16,252 +16,183 @@ limitations under the License.
-
-    sEnumerateDevicesCallback callBack;
-    callBack.Data = (void*)&Paths;
-    callBack.Function = [](void* Data, const HDEVINFO& DevInfoHandle, SP_DEVINFO_DATA& DevInfoData, SP_DEVICE_INTERFACE_DATA& DevInterfaceData, const PSP_INTERFACE_DEVICE_DETAIL_DATA& DevDetailData, U32 SizeOfDevDetailData, eErrorCode& ErrorCode)
+    const char* bus = udev_device_get_property_value(UdevDevice, "ID_BUS");
+    if(nullptr == bus)
     {
-        std::vector<String>* paths = ( std::vector<String>* )Data;
-        paths->push_back( DevDetailData->DevicePath );
-    };
+        return false;
+    }
+    if((strcmp(bus, "ata") == 0) || (strcmp(bus, "sata")  == 0))
+    {
+        return true;
+    }
+    return false;
+}
 
-    error = EnumerateDevices( callBack, InterfaceClassGUID, OnErrorBehavior );
-    return( error );
-}*/
+bool IsAtaDeviceBus( sStorageAdapterProperty StorageAdapterProperty )
+{
+    if(true == StorageAdapterProperty.AtaBus)
+    {
+        return true;
+    }
+    return false;
+}
 
-/*eErrorCode EnumerateDevices( sEnumerateDevicesCallback& Callback, const GUID* InterfaceClassGUID, eOnErrorBehavior OnErrorBehavior )
+bool InitStorageAdapterProperty( sStorageAdapterProperty& StorageAdapterProperty, udev_device* UdevDevice )
 {
-    HDEVINFO devInfoHandle =
-        SetupDiGetClassDevs
-        (
-        InterfaceClassGUID,
-        NULL,
-        NULL,
-        DIGCF_PRESENT | DIGCF_DEVICEINTERFACE
-        );
-    if ( INVALID_HANDLE_VALUE == devInfoHandle )
+    if(false == IsStorageDisk(UdevDevice))
     {
-        return( eErrorCode::Unknown );
+        return false;
     }
+    if(true == IsAtaDeviceBus(UdevDevice))
+    {
+        StorageAdapterProperty.AtaBus = true;
+        return true;
+    }
+    else
+    {
+        StorageAdapterProperty.AtaBus = false;
+        return false;
+    }
+}
 
-    DWORD error = 0;
-    DWORD memberIndex = 0;
-    while ( true == true )
+eErrorCode GetStorageDevicePaths( std::vector<String>& DevicePaths, eOnErrorBehavior OnErrorBehavior )
+{
+    struct udev *UdevObject;
+    struct udev_enumerate *Enumerate;
+    struct udev_list_entry *Devices, *dev_list_entry;
+    struct udev_device *UdevDevice;
+
+    /* Create the udev object */
+    UdevObject = udev_new();
+    if (nullptr == UdevObject)
     {
-        SP_DEVICE_INTERFACE_DATA devInterfaceData;
-        devInterfaceData.cbSize = sizeof( devInterfaceData );
-        if ( SetupDiEnumDeviceInterfaces( devInfoHandle, NULL, InterfaceClassGUID, memberIndex++, &devInterfaceData ) == FALSE )
-        {
-            error = GetLastError();
-            if ( ERROR_NO_MORE_ITEMS == error )
-            {
-                break;
-            }
-            else
-            {
-                SetupDiDestroyDeviceInfoList( devInfoHandle );
-                return( eErrorCode::Unknown );
-            }
-        }
+        throw std::runtime_error("ERROR: Can't create udev\n");
+        return eErrorCode::NullPointer;
```

Comment:
Returning NullPointer doesn't make sense. The error doesn't show the point gets failed.